[k8s] Remove privileged container requirement to support GPU isolation

romilbhardwaj opened the pull request:
Closes #2721 and #2356. We used privileged containers for FUSE mounting, but that was undesirable because a) it breaks GPU isolation and b) it's a security risk. Refer to issues for more context

We now switch to a sidecar pattern, where we expose the host's /dev/fuse to SkyPilot pods through a [daemonset](https://github.com/smarter-project/smarter-device-manager) running on each node. We use this sidecar only when FUSE mounting is requested by a task (this PR adds checks for when to do so).

Example YAML used for testing on a multi-GPU kubernetes cluster. This should show one GPU in `nvidia-smi` and mounting should succeed:
```yaml
resources:
  cloud: kubernetes
  accelerators: T4:1

file_mounts:
  /mymnt:
    source: gs://romilb-notebooks
    mode: MOUNT
  /mymnt2:
    source: s3://romil-test
    mode: MOUNT

setup: |
  echo hi

run: |
  ls -l /mymnt
  date > /mymnt/hellotest.txt
  ls -l /mymnt2
  date > /mymnt2/hellotest.txt
  nvidia-smi
```


Tested:

- [x] Code formatting: `bash format.sh`
- [x] Manual test - `sky launch` with the YAML above on a multi-GPU node in a k8s cluster. Shows only 1 GPU.
- [x] Manual test - On k8s - `sky launch` a non FUSE mount cluster, then try to `sky launch` a task requesting FUSE mounts. This fails with a clean error.
- [x] Manual test - On regular cloud - `sky launch` a non FUSE mount cluster, then try to `sky launch` a task requesting FUSE mounts. This succeeds since regular clouds do not have this limitation.
- [x] Smoke tests run - https://github.com/skypilot-org/skypilot/pull/3443#issuecomment-2065900622

concretevitamin reviewed (commented):
Thanks @romilbhardwaj. Some questions (main one is about `_requires_fuse`).

---

>  Manual test - On k8s - sky launch a non FUSE mount cluster, then try to sky launch a task requesting FUSE mounts. This fails with a clean error.

Q: Can this be relaxed in the future? E.g., if users have this back compat issue. Does `kubectl apply` a bunch of things to update the existing cluster pods + bringing up the sidecar work?

---

Since we have quite a few k8s users, shall we run k8s back compat tests too? E.g., a k8s cluster launched before this PR; after upgrading past this PR, do `sky launch / exec` on that k8s cluster with various cases.

romilbhardwaj commented on sky/resources.py:
discussed offline, resolving with no action

concretevitamin commented on sky/resources.py:
```diff
@@ -131,6 +132,11 @@ def __init__(
+            is used internally by certain cloud implementations to do additional
+            setup for FUSE mounting. This flag also safeguards against using
+            FUSE mounting on existing clusters that do not support it. If None,
+            defaults to False.
```
If we default to False, this means all previously launched and serialized Resources have this field default to False. Then let's say the same YAML used to launch the previous cluster is used in `launch -c existing`, does that mean 

https://github.com/skypilot-org/skypilot/blob/11650578d666641175d46a53843b6f36a08d1a59/sky/backends/cloud_vm_ray_backend.py#L2564-L2571

will fail?

romilbhardwaj replied:
ah good catch - showed up in backward compatibility tests. fixed by changing the deserialization to default to False when reading from old clusters.

concretevitamin replied:
Ok I see why this confused me. This code here will not be reached for the case I had in mind (non-k8s) due to less_demanding_than() call above, as inside that call we only check for k8s. I'd say add a comment on L2564 here to say "Will not be reached for non-k8s". (If it were reached, we would print this resource mismatch error.)

romilbhardwaj commented on sky/task.py:
```diff
@@ -616,6 +616,14 @@ def set_resources(
         self.resources = _with_docker_login_config(resources, self.envs)
+
+        # Evaluate if the task requires FUSE and set the requires_fuse flag
+        for _, storage_obj in self.storage_mounts.items():
+            if storage_obj.mode == storage_lib.StorageMode.MOUNT:
+                for r in self.resources:
+                    r.requires_fuse = True
+                break
```
Not sure I follow the solution here - determining the value of `r.requires_fuse` requires knowing if the task has any `MOUNT` mode storages. As a result, it must be set by the task since `Resource` is decoupled from `Task`.  Do you have an example how it can be done in the `Resource.requires_fuse` getter?

concretevitamin replied:
Ah, that's right. SG.

romilbhardwaj commented:
Tested:
- [x] Kubernetes smoke - file mounts `pytest tests/test_smoke.py::test_using_file_mounts_with_env_vars --kubernetes`
- [x] Kubernetes smoke - storage mounts `pytest tests/test_smoke.py::test_kubernetes_storage_mounts --kubernetes`
- [x] GCP smoke - file mounts `pytest tests/test_smoke.py::test_using_file_mounts_with_env_vars --gcp`
- [x] GCP smoke - storage mounts `pytest tests/test_smoke.py::test_gcp_storage_mounts_with_stop`
- [x] Kubernetes manual backward compatibility - launched cluster YAML with FUSE (in description) with master branch, switched to this branch and `sky launch` again on the same cluster - works.
- [x] Kubernetes manual backward compatibility - launched cluster YAML without FUSE with master branch, switched to this branch and `sky launch` again on the same cluster - works since the previous cluster was run in privileged mode.
- [x] GCP manual backward compatibility - launched cluster YAML with FUSE (in description) with master branch, switched to this branch and `sky launch` again on the same cluster - works.
- [x] GCP manual backward compatibility - launched cluster YAML without FUSE with master branch, switched to this branch and `sky launch` again on the same cluster - works.

concretevitamin reviewed (approved):
LGTM, thanks @romilbhardwaj.
